### DISCUSSION TITLE
Firebase Cloud Firestore for SPA

### DISCUSSION THREAD

**Discussion by arcoyk** [vercel/next.js#38656] [Help] [answered]
Hi,

I'm new to Next.js and wondering if there is a best practice tutorial for Cloud Firestore for SPA.
Yes, there are great tutorials on Firebase auth or Firestore for SSR, etc, however, I couldn't find the one for SPA.

Firestore for SPA is useful when adding a discussion feature under blog posts.

----

Where I am so far:

```javascript
import { useState, useEffect } from "react";
import FirebaseApp from "../myFirebase";
import { getFirestore, collection, getDocs } from "firebase/firestore/lite";

export async function getDataFromFirebase() {
    const db = getFirestore(FirebaseApp);
    const col = collection(db, "tweetsv1");
    const docs = await getDocs(col);
    const arr = [];
    docs.forEach((d) => {
        arr.push(d.doc());
    });
    return arr;
}

export default function Hoge() {
    const [data, setData] = useState(null);
    useEffect(() => {
        setData(getDataFromFirebase());
    }, []);
    return (
        <>
            <div>Firebase</div>
            <div>{ data }</div>
        </>
    );
}
```

As you may notice, this doesn't work because getDataFromFirebase is an async function. Does anyone know how to fix it?

**Comment by icyJoseph** [comment, marked as answer]
Hi, 

You can `.then` the result of your `getDataFromFirebase()` like:

```js
useEffect(() => {
       getDataFromFirebase().then(docs => setData(docs)); // or shorthand `.then(setData)`
 }, []);
```

This has some issues though, what happens if the component unmounts, and the promise hans't resolved yet, then when it finally does, you'll try to set state on a component that's no longer there. You need to have a cancelling mechanism. 

The most basic one would be:

```js
useEffect(() => {
    let cancel = false;
    getDataFromFirebase().then((docs) => {
        if (cancel) {
            console.log('Component unmounted');
            return;
        }
        setData(docs);
    });

    return () => {
        cancel = true;
    };
}, []);
```

I'll leave it up to you to research more into this subject matter.

---

I am a bit worried about:

```js
 docs.forEach((d) => {
        arr.push(d.doc());
    });
```

I'd have to check the firebase types, but is `d.doc()` sync or async? it might be misbehaving, so you might be pushing promises to `arr`.

Also:

```js
const arr = [];
    docs.forEach((d) => {
        arr.push(d.doc());
    });
    return docs;
```

You never do anything with `arr`, besides pushing to it. Is that intentional?

**Reply by arcoyk** [reply]
Wow, that's really helpful. Thank you. As for the last code, it was a mistake (I've just fix the original code).
